src/xmipp/libraries/data/pdb.cpp

```diff
@@ -501,18 +501,29 @@ void PDBRichPhantom::read(const FileName &fnPDB, double pseudoatoms, double thre
 			// Extract atom type and position
 			// Typical line:
 			// ATOM    909  CA  ALA A 161      58.775  31.984 111.803  1.00 34.78
-			atom.name=line.substr(12,4);
-			atom.atomType = line[13];
+			// ATOM      2  CA AALA A   1      73.796  56.531  56.644  0.50 84.78           C
+			atom.record = line.substr(0,6);
+			
+			if (const char* errmsg1 = hy36decode(5, line.substr(6,5).c_str(), 5, &atom.serial);errmsg1) {
+				throw std::invalid_argument(errmsg1);
+			}
+			atom.name = line.substr(12,4);
 			atom.altloc=line[16];
 			atom.resname=line.substr(17,3);
 			atom.chainid=line[21];
-			atom.resseq = textToInteger(line.substr(22,4));
+			
+			if (const char* errmsg2 = hy36decode(4, line.substr(22,4).c_str(), 4, &atom.resseq);errmsg2) {
+				throw std::invalid_argument(errmsg2);
+			}
 			atom.icode = line[26];
 			atom.x = textToFloat(line.substr(30,8));
 			atom.y = textToFloat(line.substr(38,8));
 			atom.z = textToFloat(line.substr(46,8));
 			atom.occupancy = textToFloat(line.substr(54,6));
 			atom.bfactor = textToFloat(line.substr(60,6));
+			atom.segment = line.substr(72,4);
+			atom.atomType = line.substr(76,2);
+			atom.charge = line.substr(78,2);
```

Comment:
I'm bit worried about all these `substr` calls. We never check for length of `line`, so this can crash at runtime (https://en.cppreference.com/w/cpp/string/basic_string/substr)

Reply:
True. We probably should check for the length of the line and include these if that data is there and otherwise leave them full of spaces.

Reply:
This is now done and reading and writing pdb files works with the pdbNmaDeform test. 

The lines now differ from the gold standard ones, but the gold standard ones are now the ones that are wrong, I think.